0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+# Pre-Shared Key
+
+The Pre-Shared Key (PSK) protocol is an end-to-end transport protocol, used by the sender and receiver of an ILP payment to decide on a condition and fulfillment for a payment. The protocol also encrypts any additional data sent along with the payment. The encryption algorithm used is AES-256-CTR.
+
+As the name suggests, PSK relies on a pre-shared secret key that is shared beforehand. In application-layer protocols like SPSP, the receiver generates the shared key and shares it with the sender over an end-to-end secure communication channel such as HTTPS. 
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. So long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key. This causes no issues unless you have some other framework on top of ILP that relies on this cryptographic proof.
+
+## Flow
+
+0. The pre-shared secret key is shared out of band.
+0. The sender creates an inner (private) PSK headers, with information on the expiry and any additional data enclosed within.
+0. The sender creates a random token value, and calculates an HMAC of the token using the pre-shared secret key as the key. This HMAC will be referred to as the payment key.
```

Comment:
I think it would be good to put a short paragraph at the top that emphasizes all of the crypto in use in this protocol. It should mention that it uses an HMAC using SHA-256 to generate the fulfillment and AES-256-CTR for encrypting application data. I would put the links in that paragraph.